File: string-to-integer-atoi.java
//TC: O(n), where n is the string length
//SC: O(1)
//Running on leetcode: yes
class Solution {
    public int myAtoi(String s) {
        if(s == null) return 0;
        s = s.trim();
        if(s.length() == 0) return 0;
        char first = s.charAt(0);
        char sign = '+';
        int limit = Integer.MAX_VALUE / 10;
        int result = 0;
        if(first == '-') {
            sign = '-';    
        }
        if(first != '+' && first != '-' && !Character.isDigit(first)) return 0;
        for(int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if(Character.isDigit(c)) {
                if(sign == '+') {
                    if(result > limit) {
                        return Integer.MAX_VALUE;
                    }
                    else if(result == limit && c - '0' > 7) {
                        return Integer.MAX_VALUE;
                    }
                }
                else if(sign == '-') {
                    if(result > limit) {
                        return Integer.MIN_VALUE;
                    }
                    else if(result == limit && c - '0' > 8) {
                        return Integer.MIN_VALUE;
                    }
                }
                result = result * 10 + c - '0';
            }
            else if(i != 0) {
                break;
            }
        }
        if(sign == '-') return -result;
        return result;
    }
}

File: String_to_Integer_atoi.py

class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.lstrip(" ")
        if not s:
            return 0
        
      
        #s = s.lstrip(" ")
        sign = 1
        start = 0
        if (s[0] != '+' and s[0] != '+') and not s[0].isdigit():
            return 0
        if s[0] == '-':
            sign = -1
            start =1
        elif s[0] == '+':
            sign = 1
            start = 1
        result = 0
        limit =  (2**31) // 10
    
        for i in range(start,len(s)):
            #print(result)
            if s[i].isdigit():
                num = int(s[i])
                if sign == 1 :
                    if result > limit:
                        return 2147483647
                    if result == limit and num >= 7:
                        return 2147483647
                else:
                    if result > limit:
                        return -2147483648
                    if result == limit and num >= 8:
                        return -2147483648
                result = result * 10 + int(num)
            elif not s[i].isdigit():
                break
                
        return result *sign                       
